module/zfs/dbuf.c

```diff
@@ -2712,15 +2712,29 @@ dmu_buf_will_clone(dmu_buf_t *db_fake, dmu_tx_t *tx)
 	 * writes and clones into this block.
 	 */
 	mutex_enter(&db->db_mtx);
+	ASSERT0(db->db_level);
+	ASSERT(db->db_blkid != DMU_BONUS_BLKID);
+	ASSERT(db->db.db_object != DMU_META_DNODE_OBJECT);
```

Comment:
In the Direct I/O PR these asserts were out of the lock.  Since they should not change during the buffer life cycle I am not sure why you moved them in now.

Reply:
Yeah, I was just being over cautious here as I noticed I was checking db state outside of the `db_mtx`. Fine with moving them out of the `db_mtx`, but I would think no harm no foul.

Reply:
Should not be critical, but I'd move them out.  First to make sure there are indeed not races related to these fields. Second, to make it more clear for reader that the fields do not need the lock.  Third, under Linux assertions stops the offending CPU, while if under lock it will also block other things attempting to take the same lock, that may block more things, and while the system is likely doomed at the end, it is often still alive enough to write the panic into the logs before reboot.

Reply:
Updated.